src/main/java/chess/domain/piece/Piece.java

```diff
@@ -0,0 +1,106 @@
+package chess.domain.piece;
+
+import chess.domain.Position;
+import chess.domain.movement.Movement;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class Piece {
+
+    private final PieceType pieceType;
+    private final Color color;
+
+    public Piece(final PieceType pieceType, final Color color) {
+        this.pieceType = pieceType;
+        this.color = color;
+    }
+
+    public boolean canMove(final Position source, final Position target,
+                           final boolean firstMove, final Map<Position, Piece> pieces) {
+        return pieceType.getMovements()
+                .stream()
+                .filter(movement -> movement.isSatisfied(color, firstMove, existEnemy(target, pieces)))
+                .map(Movement::getDirection)
+                .anyMatch(direction -> direction.canReach(source, target, findObstacle(source, target, pieces)));
+    }
+
+    private List<Position> findObstacle(final Position source, final Position target,
+                                        final Map<Position, Piece> pieces) {
+        List<Position> obstacles = pieces.entrySet().stream()
+                .filter(entry -> entry.getValue().isNotEmpty())
+                .map(Entry::getKey)
+                .collect(Collectors.toList());
+
+        removeKillableDestinationObstacle(target, pieces, obstacles);
+        addObstacleBlockedOnRankMove(source, target, pieces, obstacles);
+        return obstacles;
+    }
+
+    private void removeKillableDestinationObstacle(final Position source, final Map<Position, Piece> pieces,
+                                                   final List<Position> obstacles) {
+        if (obstacles.contains(source)
+                && pieces.getOrDefault(source, new Piece(PieceType.EMPTY, Color.NONE)).isNotSameColor(color)) {
+            obstacles.remove(source);
+        }
+    }
```

Comment:
이 메서드는 remove가 실행되었는지를 호출하는 주체는 전혀 알 수 없겠군요!

Reply:
호출 주체가 알 수 없는 것을 어떻게 호출 주체를 알 수 있게 표현할 수 있을지 감이 잡히지 않습니다 🫠

아래는 제가 해결할 수 있는 방안을 아래와 같이 작성했는데요, 제이미의 인사이트를 조금만 나눠주실 수 있을까요..? 🙏

**boolean을 통해 제거되었는지 확인**
```suggestion java
    private boolean removeKillableDestinationObstacle(final Position source, final Map<Position, Piece> pieces, final List<Position> obstacles) {
        if (obstacles.contains(source)
                && pieces.getOrDefault(source, new Piece(PieceType.EMPTY, Color.NONE)).isNotSameColor(color)) {
            obstacles.remove(source);
            return true;
        }
        return false;
    }
```

Reply:
boolean으로 반환받는 방법도 좋은 방법이죠 👍 
Collection API들이 이와 같이 구현되어 있을 거에요 😃